### PR TITLE
add support for zsod models

### PR DESCRIPTION
**Pull request by MadhuM02**
ref runs: https://ml.azure.com/experiments/id/6b656a20-e5a9-4d54-af4a-8218dbf298da?wsid=/subscriptions/dbd697c3-ef40-488f-83e6-5ad4dfb78f9b/resourceGroups/madhumaddi-rg/providers/Microsoft.MachineLearningServices/workspaces/madhumaddi-hd-onboard&tid=72f988bf-86f1-41af-91ab-2d7cd011db47

### CODE REVIEW

File: assets/training/model_management/src/azureml/model/mgmt/processors/pyfunc/vision/mmdet-od-requirements.txt [hunk 1]
```diff
@@ -5,4 +5,5 @@ openmim==0.3.9
 torch==2.0.1
 torchvision==0.15.2
 transformers==4.34.0
-accelerate==0.23.0
\ No newline at end of file
+accelerate==0.23.0
+fairscale==0.4.13
```
Comment (vadthyavath) [inline assets/training/model_management/src/azureml/model/mgmt/processors/pyfunc/vision/mmdet-od-requirements.txt:9]: would like to know the exact purpose of using both packages here!

**Review by vadthyavath — APPROVED** [review]

File: assets/training/model_management/src/azureml/model/mgmt/processors/pyfunc/vision/mmdet-od-requirements.txt (hunk 1 above)
Comment (MadhuM02) [reply assets/training/model_management/src/azureml/model/mgmt/processors/pyfunc/vision/mmdet-od-requirements.txt:9]: fairscale is used by mmd and accelerate is used by hf trainer

File: assets/training/model_management/tests/dev_conda_env.yaml [hunk 2]
```diff
@@ -6,8 +6,8 @@ dependencies:
 - pycocotools==2.0.4
 - pip:
   ## src requirements
-  - mlflow~=2.3.1
-  - mlflow-skinny~=2.3.1
+  - mlflow~=2.7.1
```
[the remark this answers is not in the document]
Comment (MadhuM02) [reply assets/training/model_management/tests/dev_conda_env.yaml:9]: this env is used for running the unit tests. wil check with import team on this

File: assets/training/model_management/tests/dev_conda_env.yaml [hunk 3]
```diff
@@ -6,8 +6,8 @@ dependencies:
...
+  - mlflow-skinny~=2.7.1
   - azureml-mlflow~=1.51.0
```
[the remark this answers is not in the document]
Comment (MadhuM02) [reply assets/training/model_management/tests/dev_conda_env.yaml:11]: this env is used for running the unit tests. wil check with import team on this

File: assets/training/model_management/src/azureml/model/mgmt/processors/pyfunc/vision/detection_predict.py [hunk 4]
```diff
@@ -166,13 +167,22 @@ def predict(self, context: mlflow.pyfunc.PythonModelContext, input_data: pd.Data
         processed_images = input_data.loc[:, [MLflowSchemaLiterals.INPUT_COLUMN_IMAGE]].apply(
             axis=1, func=process_image_pandas_series
         )
-
+        if not params:
+            params = {}
+        text_prompt = params.get("text_prompt", None)
```
Comment (novaturient95) [inline assets/training/model_management/src/azureml/model/mgmt/processors/pyfunc/vision/detection_predict.py:172]: nit:  define string literals under constants

File: assets/training/model_management/src/azureml/model/mgmt/processors/pyfunc/vision/detection_predict.py [hunk 5]
```diff
@@ -166,13 +167,22 @@ def predict(self, context: mlflow.pyfunc.PythonModelContext, input_data: pd.Data
...
+        custom_entities = params.get("custom_entities", True)
+        if not text_prompt and self.language_model:
+            raise ValueError("text prompt not sent. please send text_prompt for Launguage models")
+        classes = text_prompt.split(". ") if self.language_model else self.classes
```
Comment (novaturient95) [inline assets/training/model_management/src/azureml/model/mgmt/processors/pyfunc/vision/detection_predict.py:176]: space after `dot` is intended?

File: assets/training/model_management/src/azureml/model/mgmt/processors/pyfunc/vision/detection_predict.py [hunk 6]
```diff
@@ -166,13 +167,22 @@ def predict(self, context: mlflow.pyfunc.PythonModelContext, input_data: pd.Data
         processed_images = input_data.loc[:, [MLflowSchemaLiterals.INPUT_COLUMN_IMAGE]].apply(
             axis=1, func=process_image_pandas_series
         )
-
+        if not params:
```
Comment (novaturient95) [inline assets/training/model_management/src/azureml/model/mgmt/processors/pyfunc/vision/detection_predict.py:170]: nit: can be removed as params is already initialized to {} by default

File: assets/training/model_management/src/azureml/model/mgmt/processors/pyfunc/vision/detection_predict.py [hunk 7]
```diff
@@ -142,16 +141,18 @@ def load_context(self, context: mlflow.pyfunc.PythonModelContext) -> None:
...
+                print(f"length of classes: {len(self.classes)}")
                 print("Model loaded successfully")
             except Exception:
                 print("Failed to load the the model.")
                 raise
         else:
             raise ValueError(f"invalid task type {self._task_type}")
 
-    def predict(self, context: mlflow.pyfunc.PythonModelContext, input_data: pd.DataFrame) -> pd.DataFrame:
+    def predict(self, context: mlflow.pyfunc.PythonModelContext, input_data: pd.DataFrame, params={}) -> pd.DataFrame:
```
Comment (novaturient95) [inline assets/training/model_management/src/azureml/model/mgmt/processors/pyfunc/vision/detection_predict.py:155]: add doc string for params

Comment (novaturient95) [reply assets/training/model_management/src/azureml/model/mgmt/processors/pyfunc/vision/detection_predict.py:155]: add type

File: assets/training/model_management/src/azureml/model/mgmt/processors/pyfunc/vision/detection_predict.py [hunk 8]
```diff
@@ -142,16 +141,18 @@ def load_context(self, context: mlflow.pyfunc.PythonModelContext) -> None:
                 _map_location = "cuda" if torch.cuda.is_available() else "cpu"
                 _config = Config.fromfile(model_config_path)
                 self._model = init_detector(_config, model_weights_path, device=_map_location)
-                self.classes = self._model.dataset_meta[MMDetLiterals.CLASSES]
-
+                self.classes = self._model.dataset_meta[MMDetLiterals.CLASSES] \
+                    if MMDetLiterals.CLASSES in self._model.dataset_meta else []
+                self.language_model = hasattr(self._model, "language_model")
+                print(f"length of classes: {len(self.classes)}")
```
Comment (novaturient95) [inline assets/training/model_management/src/azureml/model/mgmt/processors/pyfunc/vision/detection_predict.py:147]: any reason logger is not used over print?

File: assets/training/model_management/src/azureml/model/mgmt/processors/pyfunc/vision/detection_predict.py [hunk 9]
```diff
@@ -69,7 +69,7 @@ def __init__(
         self._inference_detector = None
         self._task_type = task_type
 
-    def _post_process_model_results(self, batch_results: List) -> List[Dict[str, Any]]:
+    def _post_process_model_results(self, batch_results: List, classes: List) -> List[Dict[str, Any]]:
```
Comment (novaturient95) [inline assets/training/model_management/src/azureml/model/mgmt/processors/pyfunc/vision/detection_predict.py:72]: Add docstring for classes

File: assets/training/model_management/src/azureml/model/mgmt/processors/pyfunc/vision/detection_predict.py (hunk 6 above)
Comment (MadhuM02) [reply assets/training/model_management/src/azureml/model/mgmt/processors/pyfunc/vision/detection_predict.py:175]: mlflow sends params=None when use calls predict without params, and in that case its failing when trying to access the keys

File: assets/training/model_management/src/azureml/model/mgmt/processors/pyfunc/vision/detection_predict.py (hunk 5 above)
Comment (MadhuM02) [reply assets/training/model_management/src/azureml/model/mgmt/processors/pyfunc/vision/detection_predict.py:181]: yea, its needed

File: assets/training/model_management/tests/dev_conda_env.yaml (hunk 2 above)
[the remark this answers is not in the document]
Comment (MadhuM02) [reply assets/training/model_management/tests/dev_conda_env.yaml]: updated to latest in other pr

File: assets/training/model_management/tests/dev_conda_env.yaml (hunk 3 above)
[the remark this answers is not in the document]
Comment (MadhuM02) [reply assets/training/model_management/tests/dev_conda_env.yaml]: updated to latest in other pr

**Review by novaturient95 — APPROVED** [review]

**Pull request merged**